specs/CookieManagement.md

```diff
@@ -0,0 +1,505 @@
+{
+    //! [GetCookies]
+    if (m_cookieManager)
+    {
+        CHECK_FAILURE(m_cookieManager->GetCookies(
+            uri.c_str(),
+            Callback<ICoreWebView2GetCookiesCompletedHandler>(
+                [this, uri](HRESULT error_code, ICoreWebView2CookieList* list) -> HRESULT {
+                    CHECK_FAILURE(error_code);
+
+                    std::wstring result;
+                    UINT cookie_list_size;
+                    CHECK_FAILURE(list->get_Count(&cookie_list_size));
+
+                    if (cookie_list_size == 0)
+                    {
+                        result += L"No cookies found.";
+                    }
+                    else
+                    {
+                        result += std::to_wstring(cookie_list_size) + L" cookie(s) found on " +
+                                  uri + L".";
+                        result += L"\n\n[";
+                        for (int i = 0; i < cookie_list_size; ++i)
+                        {
+                            wil::com_ptr<ICoreWebView2Cookie> cookie;
+                            CHECK_FAILURE(list->GetValueAtIndex(i, &cookie));
+
+                            if (cookie.get())
+                            {
+                                result += CookieToJsonString(cookie.get());
+                                if (i != cookie_list_size - 1)
+                                {
+                                    result += L",\n";
+                                }
+                            }
+                        }
+                        result += L"]";
+                    }
+                    MessageBox(nullptr, result.c_str(), L"GetCookies Result", MB_OK);
+                    return S_OK;
+                })
+                .Get()));
+    }
+    //! [GetCookies]
+}
+```
+
+## .NET and WinRT
+
+```c#
+void AddOrUpdateCookieCmdExecuted(object target, ExecutedRoutedEventArgs e)
+{
+    CoreWebView2Cookie cookie = webView.CoreWebView2.CookieManager.CreateCookie("CookieName", "CookieValue", ".bing.com", "/");
+    cookie.SameSite = CoreWebView2CookieSameSiteKind.None;
+    webView.CoreWebView2.CookieManager.AddOrUpdateCookie(cookie);
+}
+
+async void GetCookiesCmdExecuted(object target, ExecutedRoutedEventArgs e)
+{
+    IList<CoreWebView2Cookie> cookieList = await webView.CoreWebView2.CookieManager.GetCookiesAsync("https://www.bing.com");
+    for (uint i = 0; i < cookieList.Count; ++i)
+    {
+        CoreWebView2Cookie cookie = cookieList[i];
+        Cookie systemNetCookie = cookie.ToSystemNetCookie();
+        Console.WriteLine(systemNetCookie.ToString());
+    }
+}
+
+void DeleteAllCookiesCmdExecuted(object target, ExecutedRoutedEventArgs e)
+{
+    webView.CoreWebView2.CookieManager.DeleteAllCookies();
+}
+
+void DeleteCookiesCmdExecuted(object target, ExecutedRoutedEventArgs e)
+{
+    webView.CoreWebView2.CookieManager.DeleteCookies("CookieName", "https://www.bing.com");
+}
+```

```

Comment:
These samples don't really demonstrate any scenarios. They are more like functional tests. Something more scenario-like would be

## Scenario 1

In this scenario, the host application reads contoso.com cookies from the WebView to see if the user is logged in.
```cpp
Task<bool> IsUserSignedInToContoso()
{
    CoreWebView2CookieManager cookieManager = webView.CoreWebView2.CookieManager;
    IList<CoreWebView2Cookie> cookieList =
        await webView.CoreWebView2.CookieManager.GetCookiesAsync("https://contoso.com");
    foreach (CoreWebView2Cookie cookie in cookieList)
    {
        if (cookie.Name == "is_signed_in")
        {
            return cookie.Value == "1";
        }
    }
    return false;
}
```

## Scenario 2

In this scenario, the host application sets the remembered user ID, so that when the user goes to contoso, their userid is pre-filled by the site.

```cpp
void SetRememberedUserId(string id)
{
    CoreWebView2CookieManager cookieManager = webView.CoreWebView2.CookieManager;
    CoreWebView2Cookie cookie =
        cookieManager.CreateCookieWithDetails(
            "last_userid", id, "contoso.com", "/");
    cookie.IsSecure = true; // disallow http://contoso.com/
    cookieManager.AddOrUpdateCookie(cookie);
}
```

## Scenario 3

In this scenario, the host application clears the remembered user ID because the user signed out of the host app.

```cpp
void ClearRememberedUserId()
{
    CoreWebView2CookieManager cookieManager = webView.CoreWebView2.CookieManager;
    cookieManager.DeleteCookiesWithNameAndPath("last_userid", "https://contoso.com", "/");
}
```

## Scenario 4

In this scenario, the host application clears all cookies not belonging to contoso.

```cpp
Task ClearNonContosoCookies()
{
    CoreWebView2CookieManager cookieManager = webView.CoreWebView2.CookieManager;
    IList<CoreWebView2Cookie> cookieList =
        await webView.CoreWebView2.CookieManager.GetCookiesAsync(null);
    var seen = new HashSet<string>();
    foreach (CoreWebView2Cookie cookie in cookieList)
    {
        seen.Add(cookie.Domain);
    }
    foreach (string domain in seen)
    {
        if (domain != "contoso.com") cookieManager.DeleteCookies(null, domain);
    }
}
```

## Scenario 5

In this scenario, the host application clears all cookies as part of a troubleshooter.

```cpp
void ClearCookiesTroubleshooter()
{
    CoreWebView2CookieManager cookieManager = webView.CoreWebView2.CookieManager;
    cookieManager.DeleteAllCookies();

}
```

Reply:
It's OK because it's actually a snapshot of the cookies that we are enumerating.